### DISCUSSION TITLE
The problematic single-line comment

### DISCUSSION THREAD

**Discussion by nsgnkhibdk2cls0f** [objeck/objeck-lang#90] [General]
I'm trying to create Objeck syntax highlighting rules for KDE's Kate editor. I'm stuck with the comments currently. If I make multi-line comments work, then single-line comments break. If I make single-line comments work, then multi-line comments break. It's all because Objeck's design of comments. Objeck is not special in using a single `#` char for single-line comments, Ruby does too. But Ruby has a very different multi-line comment syntax that very rarely used when Objeck take inspiration from C language family's `/*` and `*/` to be `#~` and `~#`. That caused problems as the text editor doesn't know when it's multi-line comment and when it's single-line comment. If it's only single-line comment that uses `#` char then it's easy. C language family's single-line comment started with two `/` char (`//`) so it's easy to tell when it's single-line comment when it's multi-line comment. But this design is part of the language from the beginning so it's not easy to change it. It could be Ruby encountered this problem too so they decided to use a very weird syntax for multi-line comment, so weird that it's very rarely used: `=begin` and `=end`.

p/s: I used the CSharp syntax file as the foundation to build my Objeck syntax file:

https://github.com/KDE/syntax-highlighting/blob/master/data/syntax/cs.xml

**Comment by objeck** [comment]
Thanks, I will look into this. Will you file this as an issue of type enhancement?

**Reply by ghost** [reply]
Changing the comment syntax will break backward compatibility with old code. Just because some foolish text editor can't handle it then the language has to be changed to suit that editor? I don't think so. I met the same problem with Geany that they said I have to dive into C++ code to create a new lexer for Objeck as I can't reuse the lexer for C/C++. I then abandoned Geany.

**Reply by nsgnkhibdk2cls0f** [reply]
@objeck You misunderstood me. I'm not asking you to change the language.

**Reply by nsgnkhibdk2cls0f** [reply]
@tn1997tn I have alternative options to Kate and both of them could handle Objeck's syntax for comment. I think Kate definitely able to handle it too. It's only because I don't know how to do it yet. So I'm asking the community for help, if anyone has experience with Kate. Maybe I'm better to ask on the KDE community. But I'm not a heavy user of Kate. I found it on MS Store. I'm very surprised a KDE application works on Windows so I tried it. Kate also has a very powerful integrated LSP I think could be used with the Objeck language server to provide code completion, making Kate a good IDE for Objeck as good as VSCode and Sublime (Disclaimer: I'm not tried it yet).

**Reply by objeck** [reply]
Then we misunderstood each other. I was going to add a Kate XML grammar as an enhancement. BTW - Objeck has an alternative grammar that supports C-like comments and syntax.

https://github.com/objeck/objeck-lang/tree/master/docs/syntax

**Reply by ghost** [reply]
> BTW - Objeck has an alternative grammar that supports C-like comments and syntax.

Is this already a thing or you wanted to add something like this in the future? As I tried to replace current comment syntax with C's and obc was unable to compile the source. Is there any special switches to obc needed? Btw, if it's not already a thing. I suggest please don't do this. Both syntax will be available and code are free to use either syntax the author want. It will not solve the problem with Kate unable to correctly highlight the current comment syntax, only add more problem, more syntax rules to be written.

**Reply by ghost** [reply]
The current syntax is already beautiful IMO. Took a bit from each languages, the most beautiful. Please don't make it just another dialect of C.

**Reply by nsgnkhibdk2cls0f** [reply]
I created an issue for it.

**Reply by objeck** [reply]
Thanks, I will put Kate on a VM this week

**Reply by nsgnkhibdk2cls0f** [reply]
Just to note Kate also has a Windows version. On Windows, put your objeck.xml on `%USERPROFILE%\AppData\Local\org.kde.syntax-highlighting\syntax` like this document said: https://docs.kde.org/stable5/en/kate/katepart/highlight.html

**Reply by objeck** [reply]
Thanks, I have been testing Kate on Linux VM (https://github.com/objeck/objeck-lang/blob/master/docs/syntax/kate/objeck.xml)

The issue with Kate's comments

- If a single and multi line comment starts with the same character, there must be a unique character for resolution, for example, (`//` and `/*`) or (`#~` and `##`)
- This issue also exists in Notepad++ syntax highlighting
- While not ideal, the workaround for both Kate and Notepad++ is to specify that single-line be described as `##.` Doing so does not require can changes to the language however creates _ugly_ single-line comments.

**Reply by nsgnkhibdk2cls0f** [reply]
I will go asking on the KDE forum.

**Reply by nsgnkhibdk2cls0f** [reply]
https://forum.kde.org/viewtopic.php?f=225&t=176766

**Reply by nsgnkhibdk2cls0f** [reply]
@tn1997tn This will fix your problem with Geany:

https://github.com/ScintillaOrg/lexilla/issues/121

**Reply by nsgnkhibdk2cls0f** [reply]
They allowed asking for new lexer. If we could have a C++ based Scintilla lexer it's much better than the Lua based Scintilla lexers.

**Reply by nsgnkhibdk2cls0f** [reply]
Both TextAdept and Lite XL underlying use Scintilla, but they acknowledged writing C++ based Scintilla lexer is very hard so they hacked it to provide a way to write Scintilla lexer in Lua (please note that the Lua based Scintilla lexer is specific to each editor as they have their own implementation of it).

**Reply by nsgnkhibdk2cls0f** [reply]
If the C++ based Scintilla lexer is available, it will have higher precedence than the Lua based Scintilla lexer.

**Reply by ghost** [reply]
@nsgnkhibdk2cls0f IMO you should complete the Lua based Scintilla lexer first. You can't know when the C++ based Scintilla lexer will available.

**Reply by nsgnkhibdk2cls0f** [reply]
I have done the best I can. I'm not a programmer and don't understand the regexes. I organized the keywords a bit arbitrary as I don't know how to put them in the right category of their kind. The rest should depend on @objeck to refine it.

https://github.com/objeck/objeck-lang/issues/93 https://github.com/objeck/objeck-lang/issues/95 https://github.com/objeck/objeck-lang/issues/97

**Reply by nsgnkhibdk2cls0f** [reply]
Both TextAdept and Lite XL support LSP, they could be good Objeck IDE in the future (not yet tried, though. Objeck language server requires something called Code Action as I know, some LSP implementations support it, some don't).

**Reply by objeck** [reply]
Thanks, I will take a look during breaks from GTK3 development.
